### DISCUSSION TITLE
Add JavaScript object accessor on Hash

### DISCUSSION THREAD

**Discussion by ggrossetie** [opal/opal#2205] [General]
It would be nice if Opal could provide a simple syntax to read hash entry.
With Opal < 0.8, we were using this syntax `attributes.map['source-highlighter']` but as far as I understand `map` was an internal property. With Opal > 0.8, `map` was replaced by `smap` but I believe the recommended way is to use Hash methods:

``` javascript
// var value = attributes.map['doctitle'];
var value = attributes.$fetch['doctitle'];
var value = attributes['$[]']('doctitle');
```

This is straightforward but in Ruby we can read hash entry as simple as `attributes['doctitle']`. Do you think it's possible to have the same syntax in Opal ? Basically a JavaScript object with a `doctitle` attribute.

For usability, we are planning to create an API in Asciidoctor.js to wrap Opal methods. The goal is to provide a closer API from (Asciidoctor) Ruby. If you are interested I can share my work, so you can decide if you want to integrate some parts or not :smile:

**Comment by meh** [comment]
The problem is `Hash` is waaaaaaay different from a normal JS object in behavior, you'd still have to call a method on it instead of accessing it normally.

**Comment by iliabylich** [comment]
Not sure that it's a good idea, but you can try [this](http://opalrb.org/try/?code:%25x%7B%0A%20%20Object.prototype.%24fetch%20%3D%20function%28key%29%20%7B%20return%20this%5Bkey%5D%3B%20%7D%0A%7D%0A%0Aa%20%3D%20%60%7B%20doctitle%3A%20%27value%27%20%7D%60%0A%0Aputs%20a.fetch%28%27doctitle%27%29)

**Comment by ggrossetie** [comment]
> The problem is Hash is waaaaaaay different from a normal JS object in behavior, you'd still have to call a method on it instead of accessing it normally.

We can also expose "attribute" (like `smap` or `map`) so I believe we can expose every attribute dynamically ? Having said that I'm not sure this is a good idea. What about a "public" attribute `map` ?

> Not sure that it's a good idea, but you can try this

I want to achieve the opposite. In your example you can use `fetch` method to read hash entry on a JavaScript object. I want to read hash entry like JavaScript object attributes:

``` ruby
# Ruby code
value = attributes['doctitle']
```

``` javascript
// JavaScript code
var value = attributes.doctitle; // attributes is a Ruby Hash but I can access doctitle as if attributes was a JavaScript object and doctitle its attribute
```

**Comment by iliabylich** [comment]
@Mogztter There's a conflict between reading a hash entry using a `[]` method and instance variables (which is definitely wrong). Instance variables are always raw object properties. You can't get it by design
